cmd/controller/keyregistry.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+type KeyRegistry struct {
+	client         kubernetes.Interface
+	namespace      string
+	keyPrefix      string
+	keyLabel       string
+	keysize        int
+	currentKeyName string
+	keys           map[string]*rsa.PrivateKey
+	certs          map[string]*x509.Certificate
+}
+
+func NewKeyRegistry(client kubernetes.Interface, namespace, keyPrefix, keyLabel string, keysize int) *KeyRegistry {
+	return &KeyRegistry{
+		client:    client,
+		namespace: namespace,
+		keyPrefix: keyPrefix,
+		keysize:   keysize,
+		keyLabel:  keyLabel,
+		keys:      make(map[string]*rsa.PrivateKey),
+		certs:     make(map[string]*x509.Certificate),
+	}
+}
+
+func (kr *KeyRegistry) generateKey() (string, error) {
+	key, cert, err := generatePrivateKeyAndCert(kr.keysize)
+	if err != nil {
+		return "", err
+	}
+	certs := []*x509.Certificate{cert}
+	generatedName, err := writeKey(kr.client, key, certs, kr.namespace, kr.keyLabel, kr.keyPrefix)
+	if err != nil {
+		return "", err
+	}
+	// Only store key to local store if write to k8s worked
+	kr.registerNewKey(generatedName, key, cert)
+	return generatedName, nil
+}
+
+// blacklistKey deletes a key from the local store and marks the corresponding k8s secret
+// as compromised. This effectively deletes the key from the sealedsecrets controller
+// while the key is still available to admins if need be
+func (kr *KeyRegistry) blacklistKey(keyname string) error {
+	if err := blacklistKey(kr.client, kr.namespace, keyname); err != nil {
+		return err
+	}
+	// Only delete if modifying the k8s secret succeeded
+	delete(kr.keys, keyname)
+	delete(kr.certs, keyname)
+	return nil
+}
+
+func (kr *KeyRegistry) registerNewKey(keyName string, privKey *rsa.PrivateKey, cert *x509.Certificate) {
+	kr.keys[keyName] = privKey
+	kr.certs[keyName] = cert
+	kr.currentKeyName = keyName
+}
+
+func (kr *KeyRegistry) latestKeyName() string {
+	return kr.currentKeyName
```

Comment:
I think this is "last registered", not necessarily "latest".  Perhaps when initialising the registry from existing keys, we should find the newest by `metadata.creationTimestamp`?

Reply:
I think we still need to fix this.

Afaics, we can either:
- sort by creationTimestamp (newest last) when registering existing keys during controller startup, so the last registered is the latest key
- set currentKeyName directly based on tracking the latest, either by a bulk `registerNewKeys()` method, or via a setter called only during controller startup
- we replace this function with a loop that iterates over all the keys and does that comparison every time (it will be cheap, since it's O(10) keys, and the keys are already in memory).  This will also require tracking the creationTimestamp somewhere, since we no longer have the k8s Secret objects at this point, so that's not the least-typing choice.